mlir/lib/Dialect/EmitC/IR/EmitC.cpp

```diff
@@ -698,6 +703,47 @@ LogicalResult emitc::LiteralOp::verify() {
     return emitOpError() << "value must not be empty";
   return success();
 }
+
+//===----------------------------------------------------------------------===//
+// LValueToRValueOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult emitc::LValueToRValueOp::verify() {
+  Type operandType = getOperand().getType();
+  Type resultType = getResult().getType();
+  if (!llvm::isa<emitc::LValueType>(operandType))
+    return emitOpError("operand must be a lvalue");
+  if (llvm::cast<emitc::LValueType>(operandType).getValue() != resultType)
+    return emitOpError("types must match");
+
+  Value result = getResult();
+  if (!result.hasOneUse()) {
```

Comment:
Why can this only have a single use? At this point, we read the value from memory (into registers), but then the value can be used as often as needed.

Reply:
I can't think of a way to explicitly materialize lvalue to rvalue conversions in C. Assigning to a local variable would model the read, but could have other side effects (like copying memory for structs / calling constructors in C++).

So the emitter doesn't emit any code for this op immediately. It rather forwards the name lookup of the result to it's operand. If the lvalue comes from a subscript op for example it would emit the subscript expression on every use.

Therefore the conversion op should have one use in the IR and it should immediately precede its user. (This get's more complicated when an op uses multiple values coming from conversion ops).

I'm open to ideas on how to express these conversions syntactically in C/C++ and then adapt the semantics/verifier of the operation.

Reply:
I think assigning to a local variable is the right thing to do here because it's consistent how emitc behaves everywhere.
E.g. if I do `emitc.add %a, %b : !emitc.opaque<"myclass">` then the translation will already not only call the overloaded `operator+` on my class, but also assign it to a variable (which might copy memory / call constructors).
If users don't want this, they need to use `emitc.expression`.

I have concerns about special requirements like `only single use` and must immediately precede its user that don't seem to exist in other parts of MLIR and thus will be surprising to developers.

Reply:
If users don't want the copy, they might convert their lvalue to an rvalue *reference* through `std::move`, but that is not what we need for the globals/arrays; because then the reference would still notice mutations of the underlaying object.

Reply:
I think this is the only reason why #93371 looks simpler than this PR for the scf-to-emitc lowering. If we use `lvalue_to_rvalue` as discussed here, it exactly goes in the spot where #93371 currently uses `memref.load`, but otherwise the code from #93371 should be perfectly usable here (with emitc instead of memref ops).